pkg/ebpf/probes/tc_probes.go

```diff
@@ -0,0 +1,198 @@
+func (p *tcProbes) addInterface(netIface *net.Interface) error {
+
+	// loopback devices special case (ingress only usually)
+	if netIface.Flags&net.FlagLoopback == net.FlagLoopback {
+		if _, ok := p.tcProbes[netIface]; !ok {
+			p.tcProbes[netIface] = loTcProbes[:]
+		}
+
+		return nil
+	}
+
+	if _, ok := p.tcProbes[netIface]; !ok {
+		p.tcProbes[netIface] = allTcProbes[:]
+	}
+
+	return nil
+}
+
+// delInterface deletes all default tcProbes from a given Interface
+func (p *tcProbes) delInterface(netIface *net.Interface) error {
+
+	delete(p.tcProbes, netIface)
+
+	return nil
+}
+
+// AttachToInterface attaches all tcProbes to their eBPF progs and Interfaces
+func (p *tcProbes) AttachToInterface(netIface *net.Interface) error {
+
+	p.addInterface(netIface)
+
+	for _, tcProbe := range p.tcProbes[netIface] {
+		err := tcProbe.attach(p.module, netIface)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DetachFromInterface detaches all tcProbes associated with an Interface
+func (p *tcProbes) DetachFromInterface(netIface *net.Interface) error {
+
+	for _, tcProbe := range p.tcProbes[netIface] {
+		err := tcProbe.detach()
+		if err != nil {
+			return err
+		}
+	}
+
+	p.delInterface(netIface)
+
+	return nil
+}
+
+// DetachAll detaches all existing tcProbes
+func (p *tcProbes) DetachAll() error {
+
+	for netIface := range p.tcProbes {
+		p.DetachFromInterface(netIface)
+	}
+
+	return nil
+}
+
+// DisableAutoload disables all tcProbes programs autoload
+func (p *tcProbes) DisableAutoload() error {
+
+	for _, tcProbe := range allTcProbes {
+		tcProbe.disableAutoload(p.module)
+	}
+
+	for _, tcProbe := range loTcProbes {
+		tcProbe.disableAutoload(p.module)
+	}
+
+	return nil
+}
+
+// tcProbe
+
+type tcProbe struct {
+	tcHook      *bpf.TcHook
+	attachPoint bpf.TcAttachPoint
+	programName string
+}
+
+// attach attaches a tcProbe to a given Interface
+func (p *tcProbe) attach(module *bpf.Module, netIface *net.Interface) error {
+	var err error
+
+	p.tcHook = module.TcHookInit()
+
+	err = p.tcHook.SetInterfaceByName(netIface.Name)
+	if err != nil {
+		return fmt.Errorf("failed to set tc hook interface: %v", err)
+	}
+
+	p.tcHook.SetAttachPoint(p.attachPoint)
+
+	err = p.tcHook.Create()
+	if err != nil {
+		if errno, ok := err.(syscall.Errno); ok && errno != syscall.EEXIST {
+			return fmt.Errorf("tc hook create: %v", err)
+		}
+	}
+
+	if module == nil {
+		fmt.Printf("eh null\n")
+	}
+
+	prog, _ := module.GetProgram(p.programName)
+	if prog == nil {
+		return fmt.Errorf("could not find tc program %s", p.programName)
+	}
+
+	var tcOpts bpf.TcOpts
+	tcOpts.ProgFd = int(prog.GetFd())
+	err = p.tcHook.Attach(&tcOpts)
+	if err != nil {
+		return fmt.Errorf("tc attach: %v", err)
+	}
+
+	return nil
+}
+
+// detach detaches a tcProbe from a given Interface
+func (p *tcProbe) detach() error {
+	var err error
+
+	if p.tcHook != nil {
+		err = p.tcHook.Destroy()
+	}
+
+	// TODO: check this
```

Comment:
I still need to check this:

<img width="1214" alt="image" src="https://user-images.githubusercontent.com/7395852/173734197-34ae7bfd-f847-48f4-8eb9-da706af88832.png">

I don't want to leave a hanging program/filter just because we're afraid of deleting the qdisc. There is an upstream discussion about this that Yaniv started but, still, there are many BPF programs being left hanging there if you execute tracee multiple times.

Reply:
I'm opening the following issue for this: https://github.com/aquasecurity/tracee/issues/1828 AND will continue the exact same behavior in this refactoring.